pinot-core/src/main/java/org/apache/pinot/core/operator/query/SelectionOperator.java

```diff
@@ -0,0 +1,269 @@
+      columnDataTypes[i] = DataSchema.ColumnDataType
+          .fromDataType(_expressionResultMetadata[i].getDataType(),
+              _expressionResultMetadata[i].isSingleValue());
+      columnNames[i] = expression.toString();
+      if (_expressionResultMetadata[i].hasDictionary()) {
+        _dictionaries[i] = _transformOperator.getDictionary(expression);
+      }
+    }
+    _dataSchema = new DataSchema(columnNames, columnDataTypes);
+    Comparator<Serializable[]> comparator = getStrictComparator();
+    _priorityQueue = new PriorityQueue<>(_maxRows, comparator);
+    _rowEvents = new ArrayList<>();
+  }
+
+  private Comparator<Serializable[]> getStrictComparator() {
+    return new Comparator<Serializable[]>() {
+      @Override
+      public int compare(Serializable[] o1, Serializable[] o2) {
+        List<SelectionSort> sortSequence = _selection.getSelectionSortSequence();
+        int numSortColumns = sortSequence.size();
+        for (int i = 0; i < numSortColumns; i++) {
+          int ret = 0;
+          SelectionSort selectionSort = sortSequence.get(i);
+          Serializable v1 = o1[i];
+          Serializable v2 = o2[i];
+          int index = _orderByIndices.get(i);
+          DataType dataType = _expressionResultMetadata[index].getDataType();
+          // Only compare single-value columns.
+          switch (dataType) {
+            case INT:
+              if (!selectionSort.isIsAsc()) {
+                ret = ((Integer) v1).compareTo((Integer) v2);
+              } else {
+                ret = ((Integer) v2).compareTo((Integer) v1);
+              }
+              break;
+            case LONG:
+              if (!selectionSort.isIsAsc()) {
+                ret = ((Long) v1).compareTo((Long) v2);
+              } else {
+                ret = ((Long) v2).compareTo((Long) v1);
+              }
+              break;
+            case FLOAT:
+              if (!selectionSort.isIsAsc()) {
+                ret = ((Float) v1).compareTo((Float) v2);
+              } else {
+                ret = ((Float) v2).compareTo((Float) v1);
+              }
+              break;
+            case DOUBLE:
+              if (!selectionSort.isIsAsc()) {
+                ret = ((Double) v1).compareTo((Double) v2);
+              } else {
+                ret = ((Double) v2).compareTo((Double) v1);
+              }
+              break;
+            case STRING:
+              if (!selectionSort.isIsAsc()) {
+                ret = ((String) v1).compareTo((String) v2);
+              } else {
+                ret = ((String) v2).compareTo((String) v1);
+              }
+              break;
+            default:
+              break;
+          }
+
+          if (ret != 0) {
+            return ret;
+          }
+        }
+        return 0;
+      }
+    };
+
+  }
+
+  @Override
+  protected IntermediateResultsBlock getNextBlock() {
+    int numDocsScanned = 0;
+
+    TransformBlock transformBlock;
+    if (_orderByExpressions.isEmpty()) {
```

Comment:
Would an inner 'if' be better instead of the outer 'if', where only the PQ logic is inside the 'if'? Otherwise, the code outside of 'if' gets duplicated.

Reply:
fixed